File: test/firestore_catalog_mapper_test.dart
import 'package:blindbox_app/features/catalog/firestore/firestore_catalog_mapper.dart';
import 'package:cloud_firestore/cloud_firestore.dart';
import 'package:flutter_test/flutter_test.dart';

void main() {
  group('mapFirestoreSeries', () {
    test('parses live-shaped series doc with aliases and Timestamp releaseDate', () {
      final mapped = mapFirestoreSeries(
        'aespa_fluffy_club_vinyl_plush_doll_pendant_series',
        {
          'aliases': [
            'aespa Fluffy Club Series',
            'aespa fluffy club vinyl plush doll pendant',
          ],
          'brandId': 'pop_mart',
          'displayName': 'aespa Fluffy Club Series Vinyl Plush Doll Pendant',
          'id': 'aespa_fluffy_club_vinyl_plush_doll_pendant_series',
          'imageKey': 'aespa_fluffy_club_vinyl_plush_doll_pendant_series',
          'ipId': 'aespa',
          'isBlindBox': true,
          'releaseDate': Timestamp.fromDate(DateTime.utc(2026, 4, 30)),
        },
      );
      expect(mapped, isNotNull);
      final s = mapped!;
      expect(s.id, 'aespa_fluffy_club_vinyl_plush_doll_pendant_series');
      expect(s.imageKey, s.id);
      expect(s.releaseDate, '2026-04-30');
      expect(s.aliases, hasLength(2));
    });
  });

  group('mapFirestoreFigure', () {
    test('parses live-shaped figure doc with null rarityLabel', () {
      final mapped = mapFirestoreFigure(
        'aespa_fluffy_club_vinyl_plush_doll_pendant_series_fiuffy_gelbulnyangi_giselle_ver',
        {
          'brandId': 'pop_mart',
          'displayName': 'Fluffy GELBULNYANGI(GISELLE VER.)',
          'id':
              'aespa_fluffy_club_vinyl_plush_doll_pendant_series_fiuffy_gelbulnyangi_giselle_ver',
          'imageKey':
              'aespa_fluffy_club_vinyl_plush_doll_pendant_series_fiuffy_gelbulnyangi_giselle_ver',
          'ipId': 'aespa',
          'isSecret': false,
          'rarityLabel': null,
          'seriesId': 'aespa_fluffy_club_vinyl_plush_doll_pendant_series',
          'sortOrder': 2,
        },
      );
      expect(mapped, isNotNull);
      final f = mapped!;
      expect(f.imageKey, f.id);
      expect(f.rarityLabel, isNull);
      expect(f.seriesId, 'aespa_fluffy_club_vinyl_plush_doll_pendant_series');
    });

    test('skips figure without seriesId', () {
      final mapped = mapFirestoreFigure('fig_orphan', {
        'brandId': 'pop_mart',
        'displayName': 'Orphan',
        'imageKey': 'fig_orphan',
        'ipId': 'aespa',
        'isSecret': false,
        'sortOrder': 1,
      });
      expect(mapped, isNull);
    });
  });
}

File: tools/catalog/_firestore_catalog_mapper.test.mjs
import assert from 'node:assert/strict';
import test from 'node:test';

import {
  compareSeriesSnapshotOrder,
  firestoreCatalogDocToJsonMap,
  mapFirestoreFigure,
  mapFirestoreSeries,
  mapSeriesSnapshot,
  timestampToCatalogDate,
} from './_firestore_catalog_mapper.mjs';

test('firestoreCatalogDocToJsonMap uses doc id when id missing', () => {
  const mapped = firestoreCatalogDocToJsonMap('series_a', {
    brandId: 'pop_mart',
    displayName: 'Series A',
  });
  assert.equal(mapped.id, 'series_a');
});

test('firestoreCatalogDocToJsonMap converts Firestore Timestamp releaseDate', () => {
  const mapped = firestoreCatalogDocToJsonMap('series_a', {
    id: 'series_a',
    releaseDate: {
      toDate: () => new Date('2026-04-23T15:30:00Z'),
    },
  });
  assert.equal(mapped.releaseDate, '2026-04-23');
});

test('mapFirestoreSeries rejects missing imageKey', () => {
  const mapped = mapFirestoreSeries('series_a', {
    id: 'series_a',
    brandId: 'pop_mart',
    ipId: 'labubu',
    displayName: 'Series A',
  });
  assert.equal(mapped, null);
});

test('mapFirestoreFigure keeps JSON-compatible fields', () => {
  const mapped = mapFirestoreFigure('figure_a', {
    id: 'figure_a',
    seriesId: 'series_a',
    brandId: 'pop_mart',
    ipId: 'labubu',
    displayName: 'Hope',
    imageKey: 'figure_a',
    isSecret: false,
    sortOrder: 1.0,
    rarityLabel: null,
  });

  assert.deepEqual(mapped, {
    id: 'figure_a',
    seriesId: 'series_a',
    brandId: 'pop_mart',
    ipId: 'labubu',
    displayName: 'Hope',
    imageKey: 'figure_a',
    isSecret: false,
    sortOrder: 1,
    rarityLabel: null,
    releaseDate: null,
  });
});

test('mapFirestoreFigure normalizes figure aliases', () => {
  const mapped = mapFirestoreFigure('figure_a', {
    id: 'figure_a',
    seriesId: 'series_a',
    brandId: 'pop_mart',
    ipId: 'labubu',
    displayName: 'Luck',
    imageKey: 'figure_a',
    isSecret: false,
    sortOrder: 1,
    aliases: ['Lucky', '', '  ', 'Luckster'],
  });

  assert.deepEqual(mapped?.aliases, ['Lucky', 'Luckster']);
});

test('mapSeriesSnapshot sorts by releaseDate descending', () => {
  const snap = {
    docs: [
      {
        id: 'older',
        data: () => ({
          id: 'older',
          brandId: 'b',
          ipId: 'i',
          displayName: 'Older',
          imageKey: 'older',
          releaseDate: '2025-01-01',
          isBlindBox: true,
        }),
      },
      {
        id: 'newer',
        data: () => ({
          id: 'newer',
          brandId: 'b',
          ipId: 'i',
          displayName: 'Newer',
          imageKey: 'newer',
          releaseDate: '2026-01-01',
          isBlindBox: true,
        }),
      },
    ],
  };

  const series = mapSeriesSnapshot(snap);
  assert.deepEqual(series.map((row) => row.id), ['newer', 'older']);
});

test('compareSeriesSnapshotOrder matches Dart ordering', () => {
  const newer = { id: 'newer', releaseDate: '2026-01-01' };
  const older = { id: 'older', releaseDate: '2025-01-01' };
  assert.ok(compareSeriesSnapshotOrder(newer, older, 0, 1) < 0);
  assert.ok(compareSeriesSnapshotOrder(older, newer, 1, 0) > 0);
});

test('timestampToCatalogDate formats UTC date', () => {
  assert.equal(
    timestampToCatalogDate({
      toDate: () => new Date('2026-06-15T23:59:59Z'),
    }),
    '2026-06-15',
  );
});
